cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1058,6 +1062,109 @@ void AddFSBinaryIfElseKernel(const std::shared_ptr<IfElseFunction>& scalar_funct
   DCHECK_OK(scalar_function->AddKernel(std::move(kernel)));
 }
 
+// Given a reference dictionary, computes indices to map dictionary values from a
+// comparison dictionary to the reference.
+class DictionaryRemapper {
```

Comment:
Don't we already have `DictionaryUnifier` for this? Or am I misunderstanding?

Reply:
This is for if we don't want unification, however, I think we might want to just unify dictionaries always.

IIRC, what DictionaryUnifier was missing was a way to compute a transposition map without adding new values to the internal memo table.